File: utils/handlers/inventory.js
const items = require('../data/items');
const ui = require('../ui');
const db = require('../../database');
const media = require('../media');
const { Markup } = require('telegraf');

/**
 * Inventory Handler: Displays and manages the user's collected items.
 */
module.exports = {
    async showInventory(ctx) {
        let user = await db.users.findOne({ telegramId: ctx.from.id });
        const inv = user.inventory || [];
        
        const potions = inv.filter(i => items[i.id]?.name.toLowerCase().includes('potion'));
        
        let msg = `⚡ <b>Yᴏᴜʀ Iɴᴠᴇɴᴛᴏʀʏ</b> 👇\n\n` +
            `• <b>Tɪᴄᴋᴇᴛꜱ :</b> <code>${user.gachaTickets || 0}</code>\n` +
            `• <b>Cᴜʀᴇꜱᴇᴅ Eɴᴇʀɢʏ :</b> <code>${user.stamina || 0}/100</code>\n` +
            `• <b>Pᴏᴛɪᴏɴꜱ :</b> <code>${potions.length}</code>\n\n` +
            `💰 <b>Gᴏʟᴅꜱ :</b> <code>${user.gems || 0}</code>\n` +
            `💎 <b>Sʜᴀʀᴅꜱ :</b> <code>${user.shardsCurrency || 0}</code>\n` +
            `🪙 <b>Cᴏɪɴꜱ :</b> <code>${user.coins || 0}</code>\n\n` +
            `<i>Tap an item button below to use or view details.</i>`;

        const kb = [];
        inv.forEach(entry => {
            const data = items[entry.id];
            if (data) {
                kb.push([Markup.button.callback(`${data.icon} ${data.name} (x${entry.qty})`, `view_inv_${entry.id}`)]);
            }
        });
        kb.push([Markup.button.callback('🔙 Return to Hub', 'back_to_hub')]);

        if (ctx.callbackQuery) {
            return media.smartEdit(ctx, msg, Markup.inlineKeyboard(kb));
        }
        return media.sendBanner(ctx, "inventory", msg, Markup.inlineKeyboard(kb));
    },

    async viewItemDetails(ctx, itemId) {
        const item = items[itemId];
        if (!item) return ctx.answerCbQuery("Item not found.");

        const user = await db.users.findOne({ telegramId: ctx.from.id });
        const invEntry = (user.inventory || []).find(i => i.id === itemId);
        if (!invEntry || invEntry.qty <= 0) return ctx.answerCbQuery("You no longer own this item.");

        const type = item.type || item.shop?.category || "Consumable";
        let msg = ui.formatHeader(item.name) + "\n\n" +
            `Type: <b>${type.toUpperCase()}</b>\n` +
            `Quantity: <b>${invEntry.qty}</b>\n\n` +
            `<i>${item.description}</i>`;

        const kb = [];
        // Only show USE button for specific items
        const usableOutOfBattle = ['energy_drink', 'exp_ticket', 'cursed_charm', 'exp_charm', 'minor_hp_potion', 'ce_charge'];
        if (usableOutOfBattle.includes(itemId)) {
            kb.push([Markup.button.callback(`✨ Use ${item.name}`, `use_inv_${itemId}`)]);
        }
        kb.push([Markup.button.callback('⬅️ Back to Bag', 'cmd_inv')]);

        return media.smartEdit(ctx, msg, Markup.inlineKeyboard(kb));
    },

    async useItem(ctx, itemId) {
        const user = await db.users.findOne({ telegramId: ctx.from.id });
        const inv = user.inventory || [];
        const invIdx = inv.findIndex(i => i.id === itemId);

        if (invIdx === -1 || inv[invIdx].qty <= 0) {
            return ctx.answerCbQuery("❌ You don't have enough of this item!");
        }

        let updateResult = "";
        const setOps = {};
        const incOps = {};

        if (itemId === 'energy_drink') {
            incOps.dailyExploreCount = -100; // Large reduction for drink
            updateResult = "✅ Hunt Limit Restored! (Daily hunt count reduced by 100)";
        } else if (itemId === 'max_stamina_potion') {
            setOps.dailyExploreCount = 0;
            updateResult = "🔥 EXPEDITION BURST! Daily limit reset to 0/500.";
        } else if (itemId === 'soul_shard') {
            incOps.dust = 100;
            updateResult = "✨ Soul Shard Shattered! +100 Dust acquired.";
        } else if (itemId === 'mystery_box') {
            const roll = Math.random();
            if (roll < 0.4) {
                incOps.coins = 1000;
                updateResult = "🎁 Mystery Box Opened: Found 1000 Coins!";
            } else if (roll < 0.7) {
                incOps.gachaTickets = 2;
                updateResult = "🎁 Mystery Box Opened: Found 2x Gacha Tickets!";
            } else {
                incOps.dust = 50;
                updateResult = "🎁 Mystery Box Opened: Found +50 Dust!";
            }
        } else if (itemId === 'exp_ticket') {
            const expiry = Date.now() + (24 * 60 * 60 * 1000);
            setOps.hasExplorationTicket = true;
            setOps.expTicketExpiry = expiry;
            updateResult = "✅ Activation Successful! You have unlimited explorations for the next 24h.";
        } else if (itemId === 'cursed_charm') {
            setOps.activeCursedCharm = true;
            updateResult = "🧿 Cursed Charm Active! Your next hunt will have a massive capture bonus.";
        } else if (itemId === 'exp_charm') {
            setOps.activeExpCharm = true;
            updateResult = "✨ EXP Charm Active! Your team will gain double XP in their next battle.";
        } else if (['special_grade_potion', 'minor_hp_potion', 'major_hp_potion', 'ce_core', 'ce_charge'].includes(itemId)) {
            return ctx.answerCbQuery("⚔️ This is a combat item! Use it during a duel.", { show_alert: true });
        }

        // Standard decrement logic
        inv[invIdx].qty -= 1;
        const finalInv = inv.filter(i => i.qty > 0);
        
        await db.users.update({ telegramId: ctx.from.id }, { 
            $set: { ...setOps, inventory: finalInv },
            $inc: incOps 
        });

        await ctx.answerCbQuery(updateResult, { show_alert: true });
        return this.showInventory(ctx);
    }
};

File: utils/handlers/inventory.py
import random
import time
from aiogram import Router, types, F
from aiogram.utils.keyboard import InlineKeyboardBuilder
from aiogram.filters import Command

from database import db
from utils import ui, media
from utils.data.items import ITEMS

router = Router()

async def render_inventory(callback_or_message, user, category=None):
    inv = user.get('inventory', [])
    uid = f":uid_{user['telegramId']}"
    
    if not category:
        msg = (
            "⚡ <b>Yᴏᴜʀ Iɴᴠᴇɴᴛᴏʀʏ</b> 👇\n\n"
            f"💰 <b>Cᴏɪɴꜱ :</b> <code>{user.get('coins', 0):,}</code>\n"
            f"✨ <b>Dᴜꜱᴛ :</b> <code>{user.get('dust', 0):,}</code>\n"
            f"💎 <b>Gᴇᴍꜱ :</b> <code>{user.get('gems', 0):,}</code>\n"
            f"🎟️ <b>Tɪᴄᴋᴇᴛꜱ :</b> <code>{user.get('gachaTickets', 0):,}</code>\n\n"
            "<i>Select a category below to view your items.</i>"
        )
        builder = InlineKeyboardBuilder()
        builder.row(types.InlineKeyboardButton(text="🗡️ Weapons", callback_data=f"inv_cat_weapons{uid}"))
        builder.row(types.InlineKeyboardButton(text="🎒 Items", callback_data=f"inv_cat_items{uid}"))
        builder.row(types.InlineKeyboardButton(text="🔙 Return to Hub", callback_data=f"back_to_hub{uid}"))

        if isinstance(callback_or_message, types.CallbackQuery):
            await media.edit_banner(callback_or_message.message, "inventory", msg, reply_markup=builder.as_markup())
        else:
            await media.send_banner(callback_or_message.bot, callback_or_message.chat.id, "inventory", msg, reply_markup=builder.as_markup())
        return

    weapons_ids = ['katana', 'nails', 'cloud', 'spear']
    
    filtered_inv = []
    for entry in inv:
        item_data = ITEMS.get(entry['id'])
        if not item_data: continue
        is_weapon = entry['id'] in weapons_ids
        if category == 'weapons' and is_weapon:
            filtered_inv.append((entry, item_data))
        elif category == 'items' and not is_weapon:
            filtered_inv.append((entry, item_data))

    cat_name = "Weapons" if category == 'weapons' else "Items"
    msg = f"⚡ <b>Yᴏᴜʀ Iɴᴠᴇɴᴛᴏʀʏ - {cat_name}</b> 👇\n\n<i>Tap an item to use or view details.</i>"
    
    builder = InlineKeyboardBuilder()
    if not filtered_inv:
        msg += "\n\n<i>You don't have any items in this category.</i>"
        
    for entry, item_data in filtered_inv:
        builder.row(types.InlineKeyboardButton(
            text=f"{item_data['icon']} {item_data['name']} (x{entry['qty']})", 
            callback_data=f"view_inv_{entry['id']}{uid}"
        ))
    
    builder.row(types.InlineKeyboardButton(text="🔙 Back to Categories", callback_data=f"cmd_inv{uid}"))

    if isinstance(callback_or_message, types.CallbackQuery):
        await media.edit_banner(callback_or_message.message, "inventory", msg, reply_markup=builder.as_markup())
    else:
        await media.send_banner(callback_or_message.bot, callback_or_message.chat.id, "inventory", msg, reply_markup=builder.as_markup())

@router.message(Command("inventory", "inv"))
@router.callback_query(F.data.startswith("cmd_inv"))
async def cmd_inventory(callback_or_message: types.CallbackQuery | types.Message, user: dict):
    if not user:
        if isinstance(callback_or_message, types.CallbackQuery):
            await callback_or_message.answer("❌ Please /start first.", show_alert=True)
        else:
            await callback_or_message.reply("❌ Please /start first.")
        return
    if isinstance(callback_or_message, types.CallbackQuery):
        await callback_or_message.answer()
    await render_inventory(callback_or_message, user)

@router.callback_query(F.data.startswith("inv_cat_"))
async def cmd_inventory_category(callback: types.CallbackQuery, user: dict):
    cat = callback.data.replace("inv_cat_", "").split(":")[0]
    await callback.answer()
    await render_inventory(callback, user, category=cat)

@router.callback_query(F.data.startswith("view_inv_"))
async def view_item_details(callback: types.CallbackQuery, user: dict):
    item_id = callback.data.replace("view_inv_", "").split(":")[0]
    item = ITEMS.get(item_id)
    if not item: 
        await callback.answer("Item not found.")
        return
    await callback.answer()

    inv_entry = next((i for i in user.get('inventory', []) if i['id'] == item_id), None)
    if not inv_entry or inv_entry['qty'] <= 0: 
        return await callback.answer("You no longer own this item.")

    type_str = item.get('shop', {}).get('category', 'Consumable')
    msg = (
        ui.format_header(item['name']) + "\n\n"
        f"Type: <b>{type_str.upper()}</b>\n"
        f"Quantity: <b>{inv_entry['qty']}</b>\n\n"
        f"<i>{item['description']}</i>"
    )

    uid = f":uid_{user['telegramId']}"
    builder = InlineKeyboardBuilder()
    usable_out_of_battle = [
        'energy_drink', 'exp_ticket', 'cursed_charm', 'exp_charm', 
        'minor_hp_potion', 'ce_charge', 'elixir', 'fragment', 
        'pill', 'dshard', 'reset_orb'
    ]
    if item_id in usable_out_of_battle:
        builder.row(types.InlineKeyboardButton(text=f"✨ Use {item['name']}", callback_data=f"use_inv_{item_id}{uid}"))
    
    builder.row(types.InlineKeyboardButton(text="⬅️ Back to Bag", callback_data=f"cmd_inv{uid}"))

    await media.edit_banner(callback.message, item_id, msg, reply_markup=builder.as_markup())

@router.callback_query(F.data.startswith("use_inv_"))
async def use_item(callback: types.CallbackQuery, user: dict):
    item_id = callback.data.replace("use_inv_", "")
    inv = user.get('inventory', [])
    inv_entry = next((i for i in inv if i['id'] == item_id), None)

    if not inv_entry or inv_entry['qty'] <= 0:
        return await callback.answer("❌ You don't have enough of this item!", show_alert=True)

    update_result = ""
    set_ops = {}
    inc_ops = {}

    if item_id == 'energy_drink':
        inc_ops['dailyExploreCount'] = -100
        update_result = "✅ Hunt Limit Restored! (Daily hunt count reduced by 100)"
    elif item_id == 'max_stamina_potion':
        set_ops['dailyExploreCount'] = 0
        update_result = "🔥 EXPEDITION BURST! Daily limit reset to 0/1000."
    elif item_id == 'soul_shard':
        inc_ops['dust'] = 100
        update_result = "✨ Soul Shard Shattered! +100 Dust acquired."
    elif item_id == 'mystery_box':
        roll = random.random()
        if roll < 0.4:
            inc_ops['coins'] = 1000
            update_result = "🎁 Mystery Box Opened: Found 1000 Coins!"
        elif roll < 0.7:
            inc_ops['gachaTickets'] = 2
            update_result = "🎁 Mystery Box Opened: Found 2x Gacha Tickets!"
        else:
            inc_ops['dust'] = 50
            update_result = "🎁 Mystery Box Opened: Found +50 Dust!"
    elif item_id == 'exp_ticket':
        expiry = int((time.time() + (24 * 60 * 60)) * 1000)
        set_ops['hasExplorationTicket'] = True
        set_ops['expTicketExpiry'] = expiry
        update_result = "✅ Activation Successful! You have unlimited explorations for 24h."
    elif item_id == 'cursed_charm':
        set_ops['activeCursedCharm'] = True
        update_result = "🧿 Cursed Charm Active! Your next hunt will have a massive capture bonus."
    elif item_id == 'exp_charm':
        set_ops['activeExpCharm'] = True
        update_result = "✨ EXP Charm Active! Your team will gain double XP in their next battle."
    elif item_id == 'pill':
        inc_ops['stamina'] = 50
        update_result = "💊 Energy Pill consumed! Restored 50 Stamina."
    elif item_id == 'elixir':
        set_ops['activeElixir'] = True
        update_result = "🍷 Reverse Elixir Active! Your team will start the next battle with massive HP regeneration potential."
    elif item_id == 'fragment':
        set_ops['activeFragment'] = True
        update_result = "🧿 Cursed Fragment Absorbed! Your team starts the next battle with +50 initial CE."
    elif item_id == 'dshard':
        set_ops['activeDomainShard'] = True
        update_result = "🔥 Domain Shard Resonating! Your team's Ultimate Techniques will deal +50% damage in the next battle."
    elif item_id == 'reset_orb':
        return await callback.answer("🌀 Use the Technique Reset Orb from the /upgrades menu to select a character!", show_alert=True)
    elif item_id in ['minor_hp_potion', 'ce_charge']:
        return await callback.answer("⚔️ This is a combat item! Use it during a battle.", show_alert=True)

    # Standard decrement logic
    inv_entry['qty'] -= 1
    final_inv = [i for i in inv if i['qty'] > 0]
    
    set_ops['inventory'] = final_inv
    
    await db.users.update({"telegramId": user['telegramId']}, {
        "$set": set_ops,
        "$inc": inc_ops
    })

    # Update local user object for UI
    user.update(set_ops)
    for k, v in inc_ops.items():
        user[k] = user.get(k, 0) + v

    await callback.answer(update_result, show_alert=True)
    await render_inventory(callback, user)
